sdk/servicebus/azure-servicebus/tests/async_tests/test_queues_async.py

```diff
@@ -1103,20 +1104,63 @@ async def test_queue_message_settle_through_mgmt_link_due_to_broken_receiver_lin
                 assert len(messages) == 1
                 await messages[0].complete()
 
+
     @pytest.mark.asyncio
-    async def test_async_queue_mock_no_reusing_auto_lock_renew(self):
-        class MockReceivedMessage:
-            def __init__(self):
-                self.received_timestamp_utc = utc_now()
-                self.locked_until_utc = self.received_timestamp_utc + timedelta(seconds=10)
+    async def test_async_queue_mock_auto_lock_renew_callback(self):
+        results = []
+        async def callback_mock(renewable):
+            results.append(renewable)
+
+        auto_lock_renew = AutoLockRenew()
+        auto_lock_renew.renew_period = 1 # So we can run the test fast.
+        async with auto_lock_renew: # Check that it is called when the object expires for any reason (silent renew failure)
+            message = MockReceivedMessage(prevent_renew_lock=True)
+            auto_lock_renew.register(renewable=message, on_lock_renew_failure=callback_mock)
+            await asyncio.sleep(3)
+            assert len(results) and results[-1].expired == True
 
-            async def renew_lock(self):
-                self.locked_until_utc = self.locked_until_utc + timedelta(seconds=10)
+        auto_lock_renew = AutoLockRenew()
+        auto_lock_renew.renew_period = 1
+        async with auto_lock_renew: # Check that in normal operation it does not get called
+            auto_lock_renew.register(renewable=MockReceivedMessage(), on_lock_renew_failure=callback_mock)
+            await asyncio.sleep(3)
+            assert len(results) == 1
+
+        auto_lock_renew = AutoLockRenew()
+        auto_lock_renew.renew_period = 1
+        async with auto_lock_renew: # Check that when a message is settled, it will not get called even after expiry
+            message = MockReceivedMessage(prevent_renew_lock=True)
+            auto_lock_renew.register(renewable=message, on_lock_renew_failure=callback_mock)
+            message.settled = True
+            await asyncio.sleep(3)
+            assert len(results) == 1
 
         auto_lock_renew = AutoLockRenew()
+        auto_lock_renew.renew_period = 1
+        async with auto_lock_renew: # Check that it is called when there is an overt renew failure
+            message = MockReceivedMessage(exception_on_renew_lock=True)
+            auto_lock_renew.register(renewable=message, on_lock_renew_failure=callback_mock)
+            await asyncio.sleep(3)
+            assert len(results) == 2 and results[-1].expired == True
+
+        auto_lock_renew = AutoLockRenew()
+        auto_lock_renew.renew_period = 1
+        async with auto_lock_renew: # Check that it is not called when the renewer is shutdown
+            message = MockReceivedMessage(prevent_renew_lock=True)
+            auto_lock_renew.register(renewable=message, on_lock_renew_failure=callback_mock)
+            await auto_lock_renew.shutdown()
+            await asyncio.sleep(3)
+            assert len(results) == 2
+
+
```

Comment:
This goes back to the discussion I comment previously -- are we going to check whether the receiver the message related to is alive or not.

If so, we may need to add test for that case as well.

Reply:
:) See my reply to your post.  Great minds etc.